cowbird/permissions_synchronizer.py

```diff
@@ -74,59 +74,178 @@ def __init__(self,
-
         # For each permission mapping
         for mapping in self.mapping:
             # Does the current service has some mapped permissions?
-            if permission.service_name not in mapping:
+            if res_root_key not in mapping:
                 continue
             # Does the current permission access is covered?
-            if permission.name not in mapping[permission.service_name]:
+            if permission.name not in mapping[res_root_key]:
                 continue
             # This permission is mapped : yields matches
-            for svc, mapped_perm_name in mapping.items():
+            for res, mapped_perm_name in mapping.items():
                 # Don't synch with itself
-                if svc == permission.service_name:
+                if res == res_root_key:
                     continue
                 for perm_name in mapped_perm_name:
-                    yield svc, perm_name
+                    yield res, perm_name
 
-    def sync(self, perm_operation, permission):
-        # type: (Callable[Permission], Permission) -> None
+    def _find_matching_res(self, service_name, resource_nametype_path):
+        # type: (str, str) -> (str, int)
+        """
+        Finds a resource key that matches the input resource path, in the sync_permissions config.
+        Note that it returns the longest match and only the named segments of the path are included in the length value.
+        Any tokenized segment is ignored in the length.
+
+        @param service_name Name of the service associated with the input resource.
+        @param resource_nametype_path Full resource path name, which includes the type of each segment
+                                       (ex.: /name1:type1/name2:type2)
+        """
+        service_resources = self.services[service_name]
+        # Find which resource from the config matches with the input permission's resource tree
+        # The length of a match is determined by the number of named segments matching the input resource.
+        # MULTI_TOKEN name tokens can match the related type 0 to N times. They are ignored from the match length since
+        # it is ambiguous in the case of a 0 length match.
+        # SINGLE_TOKEN name tokens can match exactly 1 time only. They are ignored from the match length since we favor
+        # a match with specific names over another with SINGLE_TOKEN generic tokens.
+        matched_res_dict = {}
+        for res_key, res_segments in service_resources.items():
+            match_len = 0
+            res_regex = "^"
+            for segment in res_segments:
+                if segment["name"] == SINGLE_TOKEN:
+                    # match any name with specific type 1 time only
+                    res_regex += rf"/\w+:{segment['type']}"
+                elif segment["name"] == MULTI_TOKEN:
+                    # match any name with specific type, 0 or more times
+                    res_regex += rf"(/\w+:{segment['type']})*"
```

Comment:
Mmhm, if a resource name or type can contain a `:`, it could indeed be problematic. Any other character that would be safe to use instead? Or maybe I can just use a double `::`?

Reply:
Its based on WPS for example that can define `identifier=ows:process` where `ows` is internally interpreted by Geoserver as a namespaced/scoped reference. So the character is enforced. I think you must also add `-` which is not included in `\w`. Other than that, I'm not aware of any other special characters to allow.